File: critter/core/src/main/kotlin/dev/morphia/critter/conventions/PropertyConvention.java
package dev.morphia.critter.conventions;

import dev.morphia.annotations.Id;
import dev.morphia.annotations.Property;
import dev.morphia.annotations.Reference;
import dev.morphia.annotations.Version;
import dev.morphia.annotations.internal.MorphiaInternal;
import dev.morphia.config.MorphiaConfig;
import dev.morphia.mapping.Mapper;
import dev.morphia.mapping.codec.pojo.PropertyModel;
import org.jetbrains.annotations.NotNull;
import org.objectweb.asm.tree.AnnotationNode;

import java.lang.annotation.Annotation;
import java.util.List;
import java.util.Map;

/**
 * @hidden
 * @morphia.internal
 * @since 3.0
 */
@MorphiaInternal
public class PropertyConvention {
    @MorphiaInternal
    public static String mappedName(MorphiaConfig config, Map<String, Annotation> annotations, String modelName) {
        Property property = (Property) annotations.get(Property.class.getName());
        Reference reference = (Reference) annotations.get(Reference.class.getName());
        Version version = (Version) annotations.get(Version.class.getName());
        Id id = (Id) annotations.get(Id.class.getName());

        String mappedName;

        if (id != null) {
            mappedName = "_id";
        } else if (property != null && !property.value().equals(Mapper.IGNORED_FIELDNAME)) {
            mappedName = property.value();
        } else if (reference != null && !reference.value().equals(Mapper.IGNORED_FIELDNAME)) {
            mappedName = reference.value();
        } else if (version != null && !version.value().equals(Mapper.IGNORED_FIELDNAME)) {
            mappedName = version.value();
        } else {
            mappedName = config.propertyNaming().apply(modelName);
        }
        return mappedName;
    }
}

File: critter/core/src/main/kotlin/dev/morphia/critter/conventions/PropertyConvention.kt
package dev.morphia.critter.conventions

import dev.morphia.annotations.Id
import dev.morphia.annotations.Property
import dev.morphia.annotations.Reference
import dev.morphia.annotations.Transient
import dev.morphia.annotations.Version
import dev.morphia.annotations.internal.MorphiaInternal
import dev.morphia.config.MorphiaConfig
import dev.morphia.mapping.Mapper

/**
 * @since 3.0
 * @hidden @morphia.internal
 */
@MorphiaInternal
object PropertyConvention {
    fun transientAnnotations(): List<Class<out Annotation>> {
        return listOf(
            Transient::class.java,
            kotlin.jvm.Transient::class.java,
            java.beans.Transient::class.java
        )
    }

    @MorphiaInternal
    fun mappedName(
        config: MorphiaConfig,
        annotations: Map<String, Annotation>,
        modelName: String
    ): String {
        val property = annotations.get(Property::class.java.getName()) as Property?
        val reference = annotations.get(Reference::class.java.getName()) as Reference?
        val version = annotations.get(Version::class.java.getName()) as Version?
        val id = annotations.get(Id::class.java.getName()) as Id?
        return when {
            id != null -> "_id"
            property != null && property.value != Mapper.IGNORED_FIELDNAME -> property.value
            reference != null && reference.value != Mapper.IGNORED_FIELDNAME -> reference.value
            version != null && version.value != Mapper.IGNORED_FIELDNAME -> version.value
            else -> config.propertyNaming().apply(modelName)
        }
    }
}
